I5011 improve collection success msg appearance

rebmullin opened the pull request:
fixes #5011 

video
https://monosnap.com/image/52gnX55LZiRElcneQSKGCGWiJAHHwN.png


I feel like we should add some transition effect here too...

bobsilverberg commented:
> I feel like we should add some transition effect here too...

I agree, we should use a css transition to make it fade out, rather than just disappear after 5 seconds.

rebmullin commented:
hey @bobsilverberg  - I've updated to use a transition. Please let me know what you think about this and some reallocation of the spacing (I've adjusted spacing here a bit b/c IMO it still looked a little odd when the space changes):

https://monosnap.com/image/mdmViit8JIIB6UPrBLlGP0s7G9dTNf.png

also - I noticed the library has updated to v2: 
https://github.com/reactjs/react-transition-group. Should I create a new issue to do an upgrade? It looks like its used in the disco pane too..

bobsilverberg commented on tests/unit/amo/components/TestCollectionManager.js:
```diff
@@ -61,6 +64,10 @@ describe(__filename, () => {
   });
 
+  afterEach(() => {
+    clock.restore();
```
If we're only using the clock in one test, is there a reason we have to do this in `afterEach`?

rebmullin replied:
sure

rebmullin replied:
altho - i don't think it would hurt having this here... I am just thinking if another tests is added with a clock..

I also see clock some other files where it appears only to be used 1 time and has this similar set up...

bobsilverberg commented on tests/unit/amo/components/TestCollectionManager.js:
```diff
@@ -708,6 +715,17 @@ describe(__filename, () => {
+
+    state = root.state();
+    expect(state.addonAddedStatus).toEqual(null);
+    root.setProps({ hasAddonBeenAdded: true });
```
I'm not sure why you're setting the state here again, and I would expect odd results because of this. Was this just a line of code left in by accident?

rebmullin replied:
that's what I thought too but it seems that I need to get the updated state (when I do it this way it works :/) 
I also had to setProps too which I thought was odd but again it worked this way. Is it possible this is happening b/c i am missing some other piece of code here?

rebmullin commented:
>I assumed we'd just use css transitions manually, but if it makes it easier that's cool.

We could but I think this library helps with transitioning when the html is removed from the DOM. 

>Regarding the version, that's odd because Greenkeeper usually keeps us up to date on out-of-date dependencies. Are you sure the new version is the latest stable version?

From what i can tell, v2 has been out for a while (almost a year), altho it does say v1 is still actively maintained

here's a screenshot of before and after (to get an idea of spacing adjustments):
![Alt text](https://monosnap.com/image/QMiCaGVLBLWPS9Qpedsdmz1Oz69bRA.png)

![Alt text](https://monosnap.com/image/ZNcxo85JJILU1nuNZuatkhznDAnxkw.png)

also - here is an updated video which i * think * might show a little better:
https://monosnap.com/file/5Wi6NTTC5DPQlouBmOOZxaIzJh5QD9#

I agree the transition seems a little fast. I did look at photon which I think suggests even faster but yeh I agree.. best to see what @pwalm thinks for this scenario

willdurand commented on src/amo/components/CollectionManager/index.js:
```diff
@@ -89,6 +93,14 @@ export class CollectionManagerBase extends React.Component<Props, State> {
       this.setState({
         addonAddedStatus: props.hasAddonBeenAdded ?
           ADDON_ADDED_STATUS_SUCCESS : null,
+      }, () => {
```
We decided not to use the `setState` second argument in https://github.com/mozilla/addons-frontend/issues/3077.

rebmullin replied:
oh? ok.. let me try another way...

willdurand commented on src/amo/components/CollectionManager/index.js:
```diff
@@ -89,6 +93,14 @@ export class CollectionManagerBase extends React.Component<Props, State> {
           ADDON_ADDED_STATUS_SUCCESS : null,
+      }, () => {
+        if (props.hasAddonBeenAdded) {
+          setTimeout(() => {
```
I think we could inject `setTimeout` as a default prop and pass a function that executes directly in the test cases.

rebmullin replied:
hey @willdurand  - do you mean something like this?

```javascript
  static defaultProps = {
    setTimeout: () => {},
  };
```

I guess I am not totally clear of the benefit of this..? let me know if i am missing something or if something like this is ok:


```javascript:
componentWillReceiveProps(props: Props) {
    ....
 
    if (hasAddonBeenAdded !== hasAddonBeenAddedNew) {
      this.setState({
        addonAddedStatus: props.hasAddonBeenAdded ?
          ADDON_ADDED_STATUS_SUCCESS : null,
      });
    }
    
    // If the message status has been set, we'll reset the message status after 5 seconds
    if (hasAddonBeenAddedNew && hasAddonBeenAddedNew !== hasAddonBeenAdded) {
      setTimeout(this.resetMessageStatus, MESSAGE_RESET_TIME);
    }
  }
```

willdurand replied:
I mean having a prop called `setTimeout` that is configured with `setTimeout` :D

In the tests, we can pass `(func) => func()` to that prop.

willdurand commented:
> Regarding the version, that's odd because Greenkeeper usually keeps us up to date on out-of-date dependencies. Are you sure the new version is the latest stable version?

I think we never upgraded by fear of breaking things, but we should...

bobsilverberg commented:
> I think we never upgraded by fear of breaking things, but we should...

Thanks @willdurand. @rebmullin could you open a separate issue for that?

pwalm commented:
Looks beautiful! Ship it!

rebmullin commented:
> ..@rebmullin could you open a separate issue for that?

sure @bobsilverberg. But how would we go about updating this if Greenkeeper didn't? is it just a manually change?

rebmullin commented:
> In the tests, we can pass (func) => func() to that prop. That way no need to deal with ticks in the tests..

hey @willdurand, I still had to use ticks so I may be missing something. Can you please take a look at my changes when you can? thanks

willdurand commented:
> hey @willdurand, I still had to use ticks so I may be missing something. Can you please take a look at my changes when you can? thanks

Sorry, here is what I had in mind. Note that we may have to rename the prop from `setTimeout` to `_setTimeout` (and to write `_setTimeout: setTimeout` in the `defaultProps`). The idea is to fully control the `setTimeout` without having to deal with clock/tick. We often do this for things like `window` or `config` for example.

```diff
diff --git a/src/amo/components/CollectionManager/index.js b/src/amo/components/CollectionManager/index.js
index 554d44fe3..31cf9c8e7 100644
--- a/src/amo/components/CollectionManager/index.js
+++ b/src/amo/components/CollectionManager/index.js
@@ -78,7 +78,7 @@ type State = {|
 
 export class CollectionManagerBase extends React.Component<Props, State> {
   static defaultProps = {
-    setTimeout: (fn) => setTimeout(fn, MESSAGE_RESET_TIME),
+    setTimeout,
   };
 
   constructor(props: Props) {
@@ -104,7 +104,7 @@ export class CollectionManagerBase extends React.Component<Props, State> {
     }
 
     if (hasAddonBeenAddedNew && hasAddonBeenAddedNew !== hasAddonBeenAdded) {
-      this.props.setTimeout(this.resetMessageStatus);
+      this.props.setTimeout(this.resetMessageStatus, MESSAGE_RESET_TIME);
     }
   }
 
diff --git a/tests/unit/amo/components/TestCollectionManager.js b/tests/unit/amo/components/TestCollectionManager.js
index ff08d05b1..b8a2d7730 100644
--- a/tests/unit/amo/components/TestCollectionManager.js
+++ b/tests/unit/amo/components/TestCollectionManager.js
@@ -44,7 +44,6 @@ const simulateAutoSearchCallback = (props = {}) => {
 };
 
 describe(__filename, () => {
-  let clock;
   let fakeRouter;
   let store;
   const apiHost = config.get('apiHost');
@@ -55,7 +54,6 @@ describe(__filename, () => {
   beforeEach(() => {
     fakeRouter = createFakeRouter();
     store = dispatchClientMetadata().store;
-    clock = sinon.useFakeTimers();
     dispatchSignInActions({
       lang,
       store,
@@ -64,10 +62,6 @@ describe(__filename, () => {
     });
   });
 
-  afterEach(() => {
-    clock.restore();
-  });
-
   const getProps = ({
     collection = createInternalCollection({
       detail: createFakeCollectionDetail(),
@@ -707,7 +701,8 @@ describe(__filename, () => {
   });
 
   it('displays a notification for 5 seconds after an add-on has been added', () => {
-    const root = render({});
+    const setTimeoutSpy = sinon.stub();
+    const root = render({ setTimeout: setTimeoutSpy });
 
     expect(root.find(Notice)).toHaveLength(0);
 
@@ -715,18 +710,15 @@ describe(__filename, () => {
 
     expect(root.find(Notice)).toHaveLength(1);
     expect(root.find(Notice).children()).toHaveText('Added to collection');
+    expect(root).toHaveState('addonAddedStatus', ADDON_ADDED_STATUS_SUCCESS);
+    sinon.assert.calledWith(setTimeoutSpy, root.instance().resetMessageStatus, 5000);
 
-    const state = root.state();
-    expect(state.addonAddedStatus).toEqual(ADDON_ADDED_STATUS_SUCCESS);
-
-    const reset = () => {
-      root.setState({ addonAddedStatus: null });
-    };
-
-    // After 5 seconds the Notice will go away.
-    root.setProps(setTimeout(reset));
+    // Simulate the setTimeout behavior.
+    root.instance().resetMessageStatus();
+    // See: https://github.com/airbnb/enzyme/blob/master/docs/guides/migration-from-2-to-3.md#for-mount-updates-are-sometimes-required-when-they-werent-before
+    root.update();
 
-    clock.tick(6000);
+    expect(root).toHaveState('addonAddedStatus', null);
     expect(root.find(Notice)).toHaveLength(0);
   });
 
```

Does that make sense?

willdurand commented on src/amo/components/CollectionManager/index.js:
```diff
@@ -72,25 +79,40 @@ type State = {|
 export class CollectionManagerBase extends React.Component<Props, State> {
+  static defaultProps = {
+    setTimeout,
+    _window: typeof window !== 'undefined' ? window : {},
```
The alternative would be to `bind` the `setTimeout` function to `window`:

```js
    setTimeout: typeof window !== 'undefined' ? window.setTimeout.bind(window) : () => {},
```

willdurand commented on src/amo/components/CollectionManager/index.js:
```diff
@@ -317,11 +345,21 @@ export class CollectionManagerBase extends React.Component<Props, State> {
-          </Notice>
-        )}
+
+        <ReactCSSTransitionGroup
```
I forgot how to use this component.. do we have to keep it all the time in the UI tree? (or can we put it in the condition where `Notice` is).

rebmullin replied:
yes - it has to always be in the DOM to work

willdurand commented on src/amo/components/CollectionManager/index.js:
```diff
@@ -317,11 +345,21 @@ export class CollectionManagerBase extends React.Component<Props, State> {
+        <ReactCSSTransitionGroup
+          className="NoticePlaceholder"
+          component="div"
+          transitionName="overlay"
```
Is it a predefined name or custom?

rebmullin replied:
No i meant to change that... thanks for catching

willdurand commented on tests/unit/amo/components/TestCollectionManager.js:
```diff
@@ -699,15 +700,27 @@ describe(__filename, () => {
     expect(root.find(Notice).children()).toHaveText('Added to collection');
+
+    expect(root).toHaveState('addonAddedStatus', ADDON_ADDED_STATUS_SUCCESS);
+    sinon.assert.calledWith(setTimeoutSpy, root.instance().resetMessageStatus, 5000);
```
I guess here we can use `MESSAGE_RESET_TIME`

willdurand commented on tests/unit/amo/components/TestCollectionManager.js:
```diff
@@ -699,15 +700,27 @@ describe(__filename, () => {
+
+    // Simulate the setTimeout behavior.
+    root.instance().resetMessageStatus();
+    // See: https://github.com/airbnb/enzyme/blob/master/docs/guides/migration-from-2-to-3.md#for-mount-updates-are-sometimes-required-when-they-werent-before
```
Maybe we can use this link with a git tag in the url: https://github.com/airbnb/enzyme/blob/enzyme%403.3.0/docs/guides/migration-from-2-to-3.md#for-mount-updates-are-sometimes-required-when-they-werent-before. That way it should always be valid.

rebmullin commented:
@willdurand, thanks for the feedback. how's it look now?
